### PR TITLE
✨AWS-library: Added interface to AWS SSM

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

This PR introduces in the AWS-library the interfacing with [AWS System Manager service (SSM)](https://aws.amazon.com/systems-manager/features/#:~:text=AWS%20Systems%20Manager%20provides%20configuration,%2C%20firewall%20settings%2C%20and%20more.)

Excerpt of AWS SSM:
```AWS Systems Manager allows you to safely automate common and repetitive IT operations and management tasks across multiple accounts and AWS Regions.```

This PR is a pre-requisite of https://github.com/ITISFoundation/osparc-simcore/pull/5923

In a nutshell, the SSM service allows to run arbitrary commands on EC2 machines when they are running. And this is needed in order to prepare EBS-based buffer machines.

## Dev-ops checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -124,11 +124,8 @@ __check_defined = \
 .PHONY: help
 
 help: ## help on rule's targets
-ifeq ($(IS_WIN),)
```
Comment (mrnicegyu11) [inline Makefile:127]: goodbye windows

File: packages/aws-library/src/aws_library/ec2/client.py [hunk 2]
```diff
@@ -274,6 +275,24 @@ async def get_instances(
...
+        try:
+            with log_context(
+                _logger,
+                logging.INFO,
+                msg=f"stopping instances {[i.id for i in instance_datas]}",
+            ):
+                await self.client.stop_instances(
+                    InstanceIds=[i.id for i in instance_datas]
+                )
+        except botocore.exceptions.ClientError as exc:
```
Comment (mrnicegyu11) [inline packages/aws-library/src/aws_library/ec2/client.py:288]: just a thought: what if only the stopping of *some* instances fails, can you handle this without getting into a weird state? ;)

File: Makefile (hunk 1 above)
Comment (pcrespov) [reply Makefile:127]: Double check the README and remove any reference to windows as well

File: packages/settings-library/src/settings_library/ssm.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+from typing import Any, ClassVar
+
+from pydantic import Field
+
+from .base import BaseCustomSettings
+
+
+class SSMSettings(BaseCustomSettings):
+    SSM_ACCESS_KEY_ID: str
```
Comment (pcrespov) [inline packages/settings-library/src/settings_library/ssm.py:9]: For secrets use SecretStr?

File: packages/settings-library/src/settings_library/ssm.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    SSM_ENDPOINT: str | None = Field(
```
Comment (pcrespov) [inline packages/settings-library/src/settings_library/ssm.py:10]: these entpoins are http? or you want to leave it open to any str?

File: packages/service-library/src/servicelib/logging_utils.py [hunk 5]
```diff
@@ -210,11 +210,11 @@ def log_decorator_wrapper(*args, **kwargs):
...
                 except:
                     # log exception if occurs in function
-                    logger_obj.error(
+                    logger_obj.log(
+                        level,
                         "Exception: %s",
-                        sys.exc_info()[1],
                         extra=extra_args,
-                        exc_info=log_traceback,
+                        stack_info=log_traceback,
```
Comment (pcrespov) [inline packages/service-library/src/servicelib/logging_utils.py:217]: :-)

**Review by mrnicegyu11 — APPROVED** [review]
Beautiful code, impressive error handling. Thanks a lot! Only one comment from my side :muscle:

File: packages/pytest-simcore/src/pytest_simcore/aws_server.py [hunk 6]
```diff
@@ -73,6 +77,42 @@ def mocked_ec2_server_envs(
...
+        SSM_SECRET_ACCESS_KEY="xxx",  # noqa: S106
+    )
+
+
+@pytest.fixture
+def mocked_ssm_server_envs(
+    mocked_ssm_server_settings: SSMSettings,
+    monkeypatch: pytest.MonkeyPatch,
+) -> EnvVarsDict:
+    changed_envs: EnvVarsDict = mocked_ssm_server_settings.dict()
```
Comment (pcrespov) [inline packages/pytest-simcore/src/pytest_simcore/aws_server.py:112]: nice. Yes, this workflow verifies everything circularly.   settings -> env vars -> capture back as settings

File: packages/aws-library/src/aws_library/ssm/_errors.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+class SSMRuntimeError(PydanticErrorMixin, RuntimeError):
+    msg_template: str = "SSM client unexpected error"
+
+
+class SSMNotConnectedError(SSMRuntimeError):
+    msg_template: str = "Cannot connect with SSM server"
+
+
+class SSMAccessError(SSMRuntimeError):
+    code = "SSM_access.error"
```
Comment (pcrespov) [inline packages/aws-library/src/aws_library/ssm/_errors.py:13]: Q: what do you use this code for?

File: packages/aws-library/src/aws_library/ec2/client.py [hunk 8]
```diff
@@ -274,6 +275,24 @@ async def get_instances(
...
+            if (
+                exc.response.get("Error", {}).get("Code", "")
+                == "InvalidInstanceID.NotFound"
```
Comment (pcrespov) [inline packages/aws-library/src/aws_library/ec2/client.py:291]: THOUGHT: I wonder if you could also keep these error codes in _errors ?

File: packages/aws-library/src/aws_library/ec2/client.py (hunk 2 above)
Comment (sanderegg) [reply packages/aws-library/src/aws_library/ec2/client.py:288]: this is only the interface code, whatever happens with these instances is the responsibility of the client code. AKA autoscaling service

File: packages/aws-library/src/aws_library/ssm/_errors.py (hunk 7 above)
Comment (sanderegg) [reply packages/aws-library/src/aws_library/ssm/_errors.py:13]: nothing

File: Makefile (hunk 1 above)
Comment (sanderegg) [reply Makefile:127]: will check. Windows runs linux anyway

File: packages/aws-library/src/aws_library/ec2/client.py (hunk 8 above)
Comment (sanderegg) [reply packages/aws-library/src/aws_library/ec2/client.py:291]: I did not refactor the EC2 client. maybe in a next PR to have them on par.

**Pull request merged**